src/OrleansProviders/Streams/Common/PooledCache/ObjectPool.cs

```diff
@@ -0,0 +1,44 @@
+﻿
+using System;
+using System.Collections.Generic;
+
+namespace Orleans.Providers.Streams.Common
+{
+    /// <summary>
+    /// Simple object pool that uses a stack to store available objects.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class ObjectPool<T> : IObjectPool<T>
+        where T : PooledResource<T>
+    {
+        private const int DefaultPoolCapacity = 1 << 10; // 1k
+        private readonly Stack<T> pool;
+        private readonly Func<IObjectPool<T>, T> factoryFunc;
+
+        public ObjectPool(Func<IObjectPool<T>, T> factoryFunc, int initialCapacity = DefaultPoolCapacity)
+        {
+            if (factoryFunc == null)
+            {
+                throw new ArgumentNullException("factoryFunc");
+            }
+            if (initialCapacity < 0)
+            {
+                throw new ArgumentOutOfRangeException("initialCapacity");
+            }
+            this.factoryFunc = factoryFunc;
+            pool = new Stack<T>(initialCapacity);
+        }
+
+        public virtual T Allocate()
+        {
+            return pool.Count != 0
+                ? pool.Pop()
+                : factoryFunc(this);
+        }
+
+        public void Free(T resource)
+        {
+            pool.Push(resource);
```

Comment:
>"I would only expect a few of them at most (maybe 5 or 10?) waiting for reallocation."

It is true that this cache mechanism, over time, may allocate more objects than are needed, but there is no indication so far that this number of unallocated blocks is significant.  In the original 343 code, we had more logging and performance counters than what has been ported.  These were not ported because the mechanisms used for this were proprietary.  In future passes, logging and performance metrics using Orleans tooling will be introduced, and if this logging indicates a need for caps, we should add them.  At this stage of development, I’m reluctant to add complexity based on speculation.

With this PR I'll open an issue to add logging and performance metrics to pooled queue cache.